### PR TITLE
add consumable base class

### PR DESCRIPTION
**Pull request by aaron42051**
I'm unsure of where we're at rn so lmk what I should do.

### CODE REVIEW

File: Assets/Character/Scripts/Constants.cs [hunk 1]
```diff
@@ -4,4 +4,10 @@
 	public const int BASE_JUMP_SPEED = 300;
     public const int BASE_ROLL_SPEED = 100;
     public const float ROLL_TIME = 1f;
+
+
+    public const int CONSUMABLE_COUNT = 3;
```
Comment (kgwong) [inline Assets/Character/Scripts/Constants.cs:9]: Each consumable type (I don't know how many we'll have, but let's use health and stamina potions as an example) will have it's own duration. So the  constants should be for each type, and not the bass class. Example: stamina potions will might have a lower cooldown than health potions.

File: Assets/Character/Scripts/Consumable.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if (Time.time - startTime > duration)
+            {
+                healing = false;
+            }
+            else
+            {
+                timer += Time.deltaTime;
+                if (timer > interval)
+                {
+                    //add hp
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable.cs]: add an abstract method for derived classes to implement. 

abstract void consume() //i don't remember the syntax

healthPotion.consume() //heals
staminaPotion.consume() //recovers stamina

File: Assets/Character/Scripts/Consumable.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                    timer = 0;
+                }
+            }
+        }
+	    else if (PlayerInput.PressedConsumable() && (Time.time - startTime > duration) && count > 0)
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable.cs]: Just expose the .consume() function mentioned above to the player. Then the player will call consume if PressedConsumble() is true. Consumable shouldn't handle any input.

File: Assets/Character/Scripts/Consumable.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class Consumable : MonoBehaviour {
+
+    private int count;
+    private float cooldown;
+    private float startTime;
+    private bool healing;
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable.cs]: what is the intention of healing?

File: Assets/Character/Scripts/Consumable.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class Consumable : MonoBehaviour {
+
+    private int count;
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable.cs]: see review summary

File: Assets/Character/Scripts/Consumable.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private float cooldown;
+    private float startTime;
+    private bool healing;
+    private float duration;
+    private float interval;
+    private float timer;
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable.cs]: in general, a bunch of primitive member vars is pretty confusing and a sign that there is a better design

**Review by kgwong — COMMENTED** [review]
I think you want a Consumable class to represent one consumable. Expose a .consume() abstract method. 

And another class ConsumableContainer<Consumable> that handles the counts and cooldowns. The player will have a list/map of ConsumableContainers for each type on consumable.

Also a bunch of .meta files were added in the commit. Not sure why.

**Comment by aaron42051** [comment]
How does the ConsumableContainer keep track of cooldowns/durations if Consumable.consume() is the class that starts the consumption? Does the Consumable class have to somehow tell the ConsumableContainer when it consumes something?

**Comment by kgwong** [comment]
ConsumableContainer will call .consume() on the consumable, if its cooldown is ready. Container will expose a function to the player.

I imagine something like the following pseudo code

if (healthPotionPressed)
{
    consumables[healthpot].use();
}

====
Container.use:

bool use()
{
   if (cooldown && count >0)
    {
          consumable.consume(...)
           count--;
           return true;
    }
    return false
}
===

File: Assets/Character/Scripts/ConsumableContainer.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class ConsumableContainer : MonoBehaviour {
+
+    Consumable c;
+    private double lastConsume;
```
Comment (kgwong) [inline Assets/Character/Scripts/ConsumableContainer.cs]: lastConsumeTime

File: Assets/Character/Scripts/Consumable1.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class Consumable1 : MonoBehaviour {
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable1.cs:4]: inherit from consumable

File: Assets/Character/Scripts/Consumable1.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            {
+                timer += Time.deltaTime;
+                if (timer > interval)
+                {
+                    //add hp
+                    timer = 0;
+                }
+            }
+        }
+	    else if (PlayerInput.PressedConsumable() && (Time.time - startTime > duration) && count > 0)
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable1.cs:40]: can you align your spaces?

File: Assets/Character/Scripts/Consumable1.cs (hunk 8 above)
Comment (kgwong) [reply Assets/Character/Scripts/Consumable1.cs:4]: i know this is an example, but a proper name would be better. HealPotionConsumable?

File: Assets/Character/Scripts/Consumable1.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class Consumable1 : MonoBehaviour {
+
+    private int count;
+    private float cooldown;
+    private float startTime;
+    private bool healing;
+    private float duration;
```
Comment (kgwong) [inline Assets/Character/Scripts/Consumable1.cs:10]: it makes more sense to define a total health and the time it takes to finish.

For example, a health potion can heal 100 hp over 10 seconds. 
Or a health potion can heal 30 hp over 2 seconds. etc etc

File: Assets/CameraController.cs.meta [hunk 11]
```diff
@@ -1,6 +1,4 @@
-fileFormatVersion: 2
-guid: 0e100b3d819732f488df1d498223e80e
-timeCreated: 1486764886
+timeCreated: 1501918168
```
Comment (kgwong) [inline Assets/CameraController.cs.meta:1]: do you know why this file was committed?

**Review by kgwong — COMMENTED** [review]
good. i suggest plugging your code into the player and see if my comments are any good :p

File: Assets/Character/Scripts/Consumable1.cs (hunk 8 above)
Comment (aaron42051) [reply Assets/Character/Scripts/Consumable1.cs:4]: ah this wasn't even intended to be an example actually, but I'll make it one

**Pull request closed without merging**